add new xscreensaver module

lasers opened the pull request:
Hello there. My best pal @neutronst4r and I aim to... Closes https://github.com/ultrabug/py3status/pull/236 


We took time today to modernize this script the best we can. One part about modernizing this script is to replace commands with `self.py3.command_run`... It works fine... kinda.

The expected behaviour for this module is to leave `xscreensaver` running in the background regardless.

If we start `xscreensaver` and later use `i3-msg start`, the process get killed or something. That was not the case with the current commands. Should we keep the current commands or to use `self.py3.command_run` with some (unknown) parameters?

Please have a look. Many thanks from us. 👨‍👦

lasers commented:
That PR is likely abandoned. Completing this PR will closes #236.

lasers commented:
I added the missing placeholder. I have been running this module for a day. Worked okay. I initially took out cache_timeout, but put it back with 60s. Waiting to be reviewed.

If we don't hear from original PR-er in a week, then we should close that PR and merge this in. I also added right-click to lock screen so this PR is a module enhancement too.

neutronst4r commented:
I am sorry for the delayed answer. I have had to study hard for finals. I have closed [#236](https://github.com/ultrabug/py3status/pull/236). And I will take a look at your proposed change.

neutronst4r commented:
Okay, I took a quick look at what you want. Using `self.py3.command_run` and having any program start with a different group id is **not** possible. (This is what you need to launch the program independent of the calling program so the called program doesn't close when the mother program is killed.) That is why I included `preexec_fn=setpgrp` in my original draft.

Without an option to pass aditional `Popen` arguments to `self.py3.command_run` it is not possible. This function could easily be added, but I don't know if it will break other modules and if the other people here even want it.

tobes commented:
> Without an option to pass aditional Popen arguments to self.py3.command_run it is not possible. 

True, but adding extra options is always possible.  Personally I'm sort of against giving complete free reign but adding a `run_independently` option that would allow this feels like maybe a way forward.

I'm quite keen to have this sort of stuff done by py3 helpers so that we can give a consistent approach eg `command_output()` returns unicode (preventing common errors) also we have an issue around Popen sometimes returning an error code of -15.

Of course this can be done later.

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,89 @@
+    cache_timeout = 60
+    format = '{output}'
+    string_off = u'🖥️'
+    string_on = u'🖥️'
```
`icon_...` would be more in keeping

also I'm not sure that this unicode char is very well supported.

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,89 @@
+    def xscreensaver(self):
+        """
+        """
+        self.run = self.py3.check_commands(['pidof xscreensaver'])
```
this is broken it is not how `check_commands()` works and you would do that in `post_config_hook()` anyhow.

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,89 @@
+        """
+        """
+        if event['button'] == 1:
+            if self.run:
```
this whole `self.run` stuff feels wrong.  we should just see if `xscreensaver` is running when we need to know - storing it just makes us more likely to have incorrect behaviour.

lasers replied:
New change. I feel I did it wrong.

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,89 @@
+        }
+
+        try:
+            self.run = check_output(['pidof', 'xscreensaver'])
```
this can still use `py3.command_output()` though you might need to change the except

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,89 @@
+Configuration parameters:
+    cache_timeout: refresh interval for this module (default 60)
+    format: default format for this module (default '{output}')
+    string_off: display when Xscreensaver is not running (defaults '🖥️')
```
default~~s~~

lasers commented:
>True, but adding extra options is always possible. Personally I'm sort of against giving complete free reign but adding a run_independently option that would allow this feels like maybe a way forward.

Should I just wait?

>this is broken it is not how check_commands() works and you would do that in post_config_hook() anyhow.

Fixed.

>icon_... would be more in keeping

Hold for pending discussion. Wanted to fix the bad code first (if any).

>also I'm not sure that this unicode char is very well supported.
FIxed. Now using `SCR` -- or you prefer `XSCR`? Blah.

>this can still use py3.command_output() though you might need to change the except

Fixed. 

I cleaned up some, but I doubt I went around with this well.

neutronst4r commented:
> Hold for pending discussion. Wanted to fix the bad code first (if any).
It really doesn't matter at all, but if you really want to use the little monitor symbol, I suggest you use [this one](http://fontawesome.io/icon/desktop/), AwesomeFont is much more widespread.

tobes commented:
>  AwesomeFont is much more widespread.

True, but I think it is important to remember that i3 is used by people for a variety of reasons and one of those is that it is lightweight.  So I think we want to be careful about adding dependencies.

@neutronst4r good to see you about again

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -11,45 +11,44 @@
-    string_off: display when Xscreensaver is not running (defaults '🖥️')
-    string_on: display when Xscreensaver is running (defaults '🖥️')
+    format_off: display when Xscreensaver is not running (default '')
+    format_on: display when Xscreensaver is running (default '')
```
`icon_..` please 

There are only so many times I can repeat myself.

lasers replied:
It was done immediately / silently. I didn't see what you said about dependencies until several hours later. `SCR`? Will look for square box.

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -11,45 +11,44 @@
     """
     """
     # available configuration parameters
-    cache_timeout = 60
```
why did you remove this? it allows user control and is pretty standard, you are now killing the cpu.

lasers replied:
Does this mean dpms should be fixed too? It's missing a `cache_timeout` too.
That's where I got the clue from.

I was trying to avoid this https://github.com/ultrabug/py3status/issues/580 by optimizing a bit. Getting it to not run something every ___ seconds, but to be updated only on click.

I think if somebody is planning on using this xscreensaver module, then they would be only xscreensaver to stop or run xscreensaver... basically CACHE_FOREVER.

It'll be back.

tobes replied:
Actually it's not updating quite as much as I feared.  But, yes we want to check.  The user may have started stopped the xscreensaver independently etc.

Thanks

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,90 @@
+        }
+
+        try:
+            self.run = self.py3.command_output(['pidof', 'xscreensaver'])
```
why set `self.run`?  there is no need to store this.  You just care about the Exception

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,90 @@
+        except:
+            self.run = False
+
+        if event['button'] == 1:
```
I quite like having `button_toggle`  `button_lock_screen` in the config then you can just do 

`if event['button'] == self.button_toggle:` etc and the user can choose the button and or disable the action. see module clock for example

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,90 @@
+                                  'xscreensaver'],
+                                 stdout=DEVNULL,
+                                 stderr=DEVNULL,
+                                 preexec_fn=setpgrp) == 0
```
This does not need `preexec_fn=setpgrp`  just use `self.py3.command_run('killall xscreensaver')` and ignore the result.

lasers commented:
@neutronst4r Thanks. I like that icon. It's much better. It fits well with the monitor from xscreensaver... https://xubuntu.org/wp-content/uploads/2014/03/xubuntu_lock_xscreensaver.png

I pushed the changes. Get with the times a bit. No more CRT monitor.

@tobes All fixed? I put `cache_timeout` back in.. w/ 15s, add `button_toggle` and `button_lock`, replaced `format_` to `icon_`. I want to keep font-awesome icon. It's more clean than `SCR` or the old icon. I hope it's ready. 👍

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,90 @@
+        """
+        """
+        try:
+            self.run = self.py3.command_output(['pidof', 'xscreensaver'])
```
can `self.run` just be `run`  by using `self.` we are storing this in the object.  Seeing as we do not use it outside of this function it is misleading and makes the code much harder to follow.  Here it is quite obvious but in a more complex module it would be nasty and leads to errors.  It is also a maintenance overhead.

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,90 @@
+                'cached_until': self.py3.time_in(self.cache_timeout),
+                'full_text': self.icon_off,
+                'color': self.py3.COLOR_BAD
+            }
```
can we just set `icon` and `color` and then create the response once.

Currently if there is no xscreensaver running we are calling `time_in()` and `safe_format()` twice which is a lot of unneeded work.  We want each module to be a light as it can be.

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,90 @@
+                self.py3.command_run(['killall', 'xscreensaver'])
+            else:
+                Popen(['xscreensaver', '-no-splash', '-no-capture-stderr'],
+                      stdout=DEVNULL, stderr=DEVNULL, preexec_fn=setpgrp) == 1
```
can we add a comment like
```
# Because we want xscreensaver to continue running even if py3status exits
# we need to run using preexec_fn=setpgrp
```

we don't need the ` == 1` test I think

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,90 @@
+
+        if event['button'] == self.button_lock:
+            Popen(['xscreensaver-command', '-lock'],
+                  stdout=DEVNULL, stderr=DEVNULL, preexec_fn=setpgrp) == 1
```
Three things.

1) this does not need preexec_fn=setpgrp so we can use `py3.command_run()`

2) we don't need the `== `1 test

3) this only works if xsreensaver is running?  So should we not check that and start xscreensaver if needed and then lock it?

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,90 @@
+
+        if event['button'] == self.button_toggle:
+            if self.run:
+                self.py3.command_run(['killall', 'xscreensaver'])
```
whilst reading the `xscreensaver-command` man page I saw this.

>-exit

>Causes the xscreensaver process to exit gracefully. This does nothing if the display is currently locked.
Warning: never use kill -9 with xscreensaver while the screensaver is active. If you are using a virtual root window manager, that can leave things in an inconsistent state, and you may need to restart your window manager to repair the damage. 

So maybe we should use this recommended method.

lasers commented:
> The review that just keeps giving

Thank you. Reviews matters.  I know it can be annoying if I made mistakes or got things done differently than you prefer. I wanted to get things done fast whilst keeping variables simple (but more importantly, things are being done correctly too).

Other issues. I will look at them and make changes in few hours.
I mark Janurary 26 as Review Appreciation Day. Cheers. 🏆

lasers commented:
> can we just set icon and color and then create the response once.

Fixed.

> Currently if there is no xscreensaver running we are calling time_in() and safe_format() twice which is a lot of unneeded work. We want each module to be a light as it can be.

Not sure what I should do here.

> can self.run just be run by using self.

Fixed. I followed dpms.


> this only works if xsreensaver is running? So should we not check that and start xscreensaver if needed and then lock it?

No running = No locking. Run it first (imho).

> -exit // Causes the xscreensaver process to exit gracefully

I tried that before. Not working. Working now tho. Maybe I used wrong py3.command. :confused:

> can we add a comment like

Yes sir. Done.

tobes commented:
> No running = No locking. Run it first (imho).

I'd say so :)

neutronst4r commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,89 @@
+ Format placeholders:
+    {icon} display current xscreensaver icon
+
+@author neutronst4r <c7420{at}gmx{dot}net>
```
change email to `<c7420{at}posteo{dot}net>` please and add yourself as an author, if you like

neutronst4r commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,89 @@
+            run = self.py3.command_output(['pidof', 'xscreensaver'])
+        except:
+            run = False
+
```
As @tobes has suggested saving the result of run is kind of unnecessary. We could do something like:
```
def _is_running_(self):
       """
       tests whether xscreensaver is running, returns boolean
       """
       try:
           self.py3.command_output(['pidof', 'xscreensaver'])
       except:
           return False
       else:
           return True
```
and ask if xscreensaver is running.

lasers replied:
Thanks. I didn't understand that. I jumped in here to muck together a wanted feature for `arch_updates` to stay hidden if none... then I keep going with many other little things.

I fixed your email and am not worthy of a mention on any script. I was wondering how to make a function so I don't do two same things in different places. This has been updated now. Cheers.

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,91 @@
+                'cached_until': self.py3.CACHE_FOREVER,
+                'color': self.py3.COLOR_BAD,
+                'full_text': STRING_UNAVAILABLE
+            }
```
This should be done in `post_config_hook()` seeing as we have it defined.  By raising an `Exception(msg)` less lines too

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,91 @@
+    def on_click(self, event):
+        run = self._is_running()
+        if event['button'] == self.button_lock:
+            self.py3.command_run(['xscreensaver-command', '-lock'])
```
should we not use `xscreensaver-command -activate`?  If I have setup my xscreensaver to not lock the screen then I would not expect it to be locked here.

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,91 @@
+ Format placeholders:
+    {icon} Xscreensaver icon
+
+@author neutronst4r <c7420{at}posteo{dot}net>
```
I don't think you should add someone else as the author also if this is based on dpms then neutronst4r is not the author of that module.

tobes replied:
actually read your initial comment here so maybe
`@author neutronst4r, lasers`

lasers replied:
Blah. Done. Let's tag it ready to go and get out of here.

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,91 @@
+"""
+from os import setpgrp
+from subprocess import Popen, DEVNULL
+STRING_UNAVAILABLE = "xscreensaver: isn't installed"
```
A blank line would be nice here

lasers commented:
Hi there.
* You want blank lines? You got them.
* Exception now in `psc` using `raise Exception(msg)` to use that new/nice exception string in py3status instead of old annoying i3-nagbar exceptions.
* I tested `-activate`. Yes. Good command. Can configure `-lock` using `xscreensaver-demo` (graphical) or edit the configuration file `~/.xscreensaver`.

tobes commented on py3status/modules/xscreensaver.py:
```diff
@@ -0,0 +1,90 @@
+    """
+    """
+    # available configuration parameters
+    button_lock = 3
```
maybe we should rename this to `button_activate` to match it's function

lasers replied:
Yup. Done.

tobes commented:
Thanks @lasers this is now good to go as far as I'm concerned.

ultrabug commented:
Okay I guess the label is good, thanks to everyone !
